pkg/oc/cli/cmd/image/mirror/mirror.go

```diff
@@ -0,0 +1,654 @@
+				continue
+			}
+
+			// filter or load manifest list as appropriate
+			srcManifests, srcDigest, err := processManifestList(ctx, srcDigest, srcManifest, manifests, src.ref, o.includeDescriptor)
+			if err != nil {
+				digestErrs = append(digestErrs, retrieverError{src: src.ref, err: err})
+				continue
+			}
+			if len(srcManifests) == 0 {
+				fmt.Fprintf(o.ErrOut, "info: Filtered all images from %s, skipping\n", src.ref)
+				continue
+			}
+
+			for _, dst := range pushTargets {
+				// if we are going to be using cross repository mount, get a token that covers the src
+				if src.ref.Registry == dst.ref.Registry {
+					toContext = toContext.WithScopes(auth.RepositoryScope{Repository: src.ref.RepositoryName(), Actions: []string{"pull"}})
+				}
+
+				toRepo, err := o.Repository(ctx, toContext, creds, dst.t, dst.ref)
+				if err != nil {
+					digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unable to connect to %s: %v", dst.ref, err)})
+					continue
+				}
+
+				canonicalTo := toRepo.Named()
+				toManifests, err := toRepo.Manifests(ctx)
+				if err != nil {
+					digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unable to access destination image %s manifests: %v", src.ref, err)})
+					continue
+				}
+
+				// if the destination tag already has this manifest, do nothing
+				var mustCopyLayers bool
+				if o.Force {
+					mustCopyLayers = true
+				} else {
+					if _, err := toManifests.Get(ctx, srcDigest); err != nil {
+						mustCopyLayers = true
+					} else {
+						glog.V(4).Infof("Manifest exists in %s, no need to copy layers without --force", dst.ref)
+					}
+				}
+				if mustCopyLayers {
+					// upload all the blobs
+					toBlobs := toRepo.Blobs(ctx)
+					srcBlobs := srcRepo.Blobs(ctx)
+
+					// upload the each manifest
+					for _, srcManifest := range srcManifests {
+						switch srcManifest.(type) {
+						case *schema2.DeserializedManifest:
+						default:
+							digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("the manifest type %T is not supported", srcManifest)})
+							continue
+						}
+
+						for _, blob := range srcManifest.References() {
+							// tagging within the same registry is a no-op
+							if src.ref.Registry == dst.ref.Registry && canonicalFrom.String() == canonicalTo.String() {
+								continue
+							}
+
+							blobSource, err := reference.WithDigest(canonicalFrom, blob.Digest)
+							if err != nil {
+								digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unexpected error building named digest: %v", err)})
+								continue
+							}
+
+							var options []distribution.BlobCreateOption
+							if !o.SkipMount {
+								options = append(options, client.WithMountFrom(blobSource), WithDescriptor(blob))
+							}
+							w, err := toBlobs.Create(ctx, options...)
```

Comment:
Some blobs may already exist in the destination repository. Doing Stat before upload would reduce traffic.

Reply:
That is happening implicitly here - uploads for blobs that already exist exits early.

Reply:
Specifically registry/client performs a POST to get the blob upload ID, at which point the registry tells us we either mounted successfully, already exists, or it is ready for us to upload.  So stat is bundled inside that check.

Reply:
When SkipMount is true, Create doesn't even get the digest, so it can't check existence of the blob.

Reply:
Create blob always POSTs:

```
func (bs *blobs) Create(ctx context.Context, options ...distribution.BlobCreateOption) (distribution.BlobWriter, error) {
	var opts distribution.CreateOptions

	for _, option := range options {
		err := option.Apply(&opts)
		if err != nil {
			return nil, err
		}
	}

	var values []url.Values

	if opts.Mount.ShouldMount {
		values = append(values, url.Values{"from": {opts.Mount.From.Name()}, "mount": {opts.Mount.From.Digest().String()}})
	}

	u, err := bs.ub.BuildBlobUploadURL(bs.name, values...)
	if err != nil {
		return nil, err
	}

	resp, err := bs.client.Post(u, "", nil)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
```

Reply:
Yes, POST can initiate a new upload or cross-mount a blob from another repository. But unfortunately it doesn't check if the blob is already here. Without the ShouldMount option, Create doesn't even know which blob is going to be uploaded.

Reply:
Fixed, thanks for clarifying.  `--force` will bypass the stat check.